homeassistant/components/husqvarna_automower/sensor.py

```diff
@@ -396,19 +401,63 @@ class AutomowerSensorEntityDescription(SensorEntityDescription):
 )
 
 
+@dataclass(frozen=True, kw_only=True)
+class WorkAreaSensorEntityDescription(SensorEntityDescription):
+    """Describes the work area sensor entities."""
+
+    exists_fn: Callable[[WorkArea], bool] = lambda _: True
+    value_fn: Callable[[WorkArea], int | datetime | None]
+    translation_key_fn: Callable[[int, str], str]
+
+
+WORK_AREA_SENSOR_TYPES: tuple[WorkAreaSensorEntityDescription, ...] = (
+    WorkAreaSensorEntityDescription(
+        key="progress",
+        translation_key_fn=_work_area_translation_key,
+        exists_fn=lambda data: data.progress is not None,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda data: data.progress,
+    ),
+    WorkAreaSensorEntityDescription(
+        key="last_time_completed",
+        translation_key_fn=_work_area_translation_key,
+        exists_fn=lambda data: data.last_time_completed_naive is not None,
+        device_class=SensorDeviceClass.TIMESTAMP,
+        value_fn=lambda data: naive_to_aware(
+            data.last_time_completed_naive,
+            ZoneInfo(str(dt_util.DEFAULT_TIME_ZONE)),
+        ),
+    ),
+)
+
+
 async def async_setup_entry(
     hass: HomeAssistant,
     entry: AutomowerConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up sensor platform."""
     coordinator = entry.runtime_data
-    async_add_entities(
-        AutomowerSensorEntity(mower_id, coordinator, description)
-        for mower_id in coordinator.data
-        for description in SENSOR_TYPES
-        if description.exists_fn(coordinator.data[mower_id])
-    )
+    entities: list[SensorEntity] = []
+    for mower_id in coordinator.data:
+        if coordinator.data[mower_id].capabilities.work_areas:
+            _work_areas = coordinator.data[mower_id].work_areas
+            if _work_areas is not None:
+                entities.extend(
+                    WorkAreaSensorEntity(
+                        mower_id, coordinator, description, work_area_id
+                    )
+                    for description in WORK_AREA_SENSOR_TYPES
+                    for work_area_id in _work_areas
+                    if description.exists_fn(_work_areas[work_area_id])
+                )
+        entities.extend(
+            AutomowerSensorEntity(mower_id, coordinator, description)
```

Comment:
(just general note, if I understand correctly, workareas are dynamic, I think it would be nice if we can subscribe to the coordinator updates here and check if there are new work areas or ones that can be removed and make sure we create and remove entities on runtime, this can also work for the lawnmower itself, but this has more gain in the first place)

Reply:
Yes, they are dynamic. Is it the same you described here? 
https://github.com/home-assistant/core/pull/126376#discussion_r1770589033

Reply:
Oh lol, I already told you apparently, yes it is

Reply:
Okay, I will have a look at it.

Reply:
But for a followup, this PR was waiting for the statetype thingy